ocaml/typing/typedecl.ml

```diff
@@ -435,24 +463,37 @@ let transl_labels ~new_var_jkind env univars closed lbls =
       lbls in
   lbls, lbls'
 
-let transl_types_gf ~new_var_jkind env univars closed tyl =
-  let mk arg =
+let transl_types_gf ~new_var_jkind ~allow_unboxed
+  env loc univars closed tyl kloc =
+  let mk idx arg =
     let cty = transl_simple_type ~new_var_jkind env ?univars ~closed Mode.Alloc.Const.legacy arg in
     let gf = Typemode.transl_global_flags
       (Jane_syntax.Mode_expr.of_attrs arg.ptyp_attributes |> fst) in
+    check_representable ~why:(Constructor_declaration idx) ~allow_unboxed
+      env loc kloc cty.ctyp_type;
     (cty, gf)
   in
-  let tyl_gfl = List.map mk tyl in
+  let tyl_gfl = List.mapi mk tyl in
   let tyl_gfl' = List.map (fun (cty, gf) -> cty.ctyp_type, gf) tyl_gfl in
   tyl_gfl, tyl_gfl'
 
-let transl_constructor_arguments ~new_var_jkind env univars closed = function
+let transl_constructor_arguments ~new_var_jkind ~unboxed
+  env loc univars closed = function
   | Pcstr_tuple l ->
-      let flds, flds' = transl_types_gf ~new_var_jkind env univars closed l in
-      Types.Cstr_tuple flds',
-      Cstr_tuple flds
+      let flds, flds' =
+        (* CR layouts: we forbid [@@unboxed] variants from being
+           non-value, see comment in [check_representable]. *)
+        transl_types_gf ~new_var_jkind ~allow_unboxed:(not unboxed)
+        env loc univars closed l (Cstr_tuple { unboxed })
```

Comment:
Indent this line a bit, just to make it clear that it's not a new function call.